---
title: "Gene Expression Network Analysis"
author: Christof Seiler
date: November, 2016
output: html_document
params:
  fdr: 0.05
  num_nodes: 100
  csv_file_name: "gestage_enroll_H1N1_results.csv"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This Rmd file contains a minimal network analysis pipeline using the packages ``STRINGdb`` and ``BioNet``.

## Install Packages

Install necessary packages from bioconductor repository. Run this code only once to install packages.

```{r eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("STRINGdb")
biocLite("BioNet")
biocLite("KEGGREST")
install.packages(c("stringr","magrittr","igraph","visNetwork","fdrtool"))
```

Load packages.

```{r warning=FALSE,message=FALSE}
library(STRINGdb)
library(BioNet)
library(DLBCL)
library(org.Hs.eg.db)
library(KEGGREST)
library(stringr)
library(magrittr)
library(igraph)
library(visNetwork)
n_cores = as.integer(Sys.getenv("SLURM_NTASKS"))
n_cores
```

## Load Data

Remove genes with NA values. As indicated in the [DESeq2 documentation](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf), some values in the results table can be set to NA for one of the following reasons:

1. If within a row, all samples have zero counts, the baseMean column will be zero, and the log2 fold change estimates, p-value and adjusted p-value will all be set to NA.
2. If a row contains a sample with an extreme count outlier then the p-value and adjusted p-value will be set to NA. These outlier counts are detected by Cook's distance.
3. If a row is filtered by automatic independent filtering, for having a low mean normalized count, then only the adjusted p-value will be set to NA.

```{r}
params$csv_file_name
diff_exp = read.csv(params$csv_file_name)
dim(diff_exp)
diff_exp = na.omit(diff_exp)
dim(diff_exp)
diff_exp$X = as.character(diff_exp$X)
diff_exp$symbol = as.character(diff_exp$symbol)
```

Remove transcipts that were not mapped to any gene symbol.

```{r}
diff_exp = subset(diff_exp,symbol != "No_Symbol_Found")
```

There are some multiple transcript to gene matches resulting in rows with the same gene sybmol. We follow a conservative approach and replace redundant genes by their maximum $p$-value.

```{r}
table(diff_exp$symbol) %>% sort(., decreasing = TRUE) %>% head
gene_names = table(diff_exp$symbol) %>% names
diff_exp = lapply(gene_names,function(gene_name) {
  duplicated = diff_exp[which(diff_exp$symbol == gene_name),]
  duplicated[which.max(duplicated$P.Value),]
}) %>% do.call(rbind,.)
```

## Download Gene Network

Download proteins for human species (code is 9606).

Consider interaciton that are 0.9 confidence. From the STRING website: 

In STRING, each protein-protein interaction is annotated with one or more 'scores'. Importantly, these scores do not indicate the strength or the specificity of the interaction. Instead, they are indicators of confidence, i.e. how likely STRING judges an interaction to be true, given the available evidence. All scores rank from 0 to 1, with 1 being the highest possible confidence. A score of 0.5 would indicate that roughly every second interaction might be erroneous (i.e., a false positive).

```{r}
string_db = STRINGdb$new(version="10", species=9606, score_threshold=900, input_directory="")
```

Check how many proteins are in the database.

```{r}
string_proteins = string_db$get_proteins()
dim(string_proteins)
```

Map gene names to identifiers used in the database.

```{r}
mapped = string_db$map(diff_exp, "symbol", removeUnmappedRows = FALSE)
interactions = string_db$get_interactions(mapped$STRING_id)
interactions = data.frame(from=interactions$from,to=interactions$to,combined_score=interactions$combined_score)
dim(interactions)
head(interactions)
```

## Find Subgraph

Fit a Beta-Uniform model.

```{r fig.width=12}
pval = mapped$P.Value
names(pval) = mapped$STRING_id
fb = fitBumModel(pval)
fb
```

Set the ``fdr`` parameter which can be interpreted as the FDR of the subgraph. Smaller values will produce a smaller maximum subgraph. You should try a few values (e.g. 0.05, 0.01, 0.001, ...) to obtain a reasonable small subgraph that permits biological interpretation.

First we make convert the interaction table into an igraph object and make the nodes names human readible.

```{r}
network = graph_from_data_frame(interactions)
```

Then we search for the optimal subgraph.

```{r}
fdr_largest = as.numeric(params$fdr)
fdr_vec = fdr_largest/(10^(0:(2*n_cores-1)))
module_list = mclapply(fdr_vec,function(fdr) {
  scores = scoreNodes(network, fb, fdr=fdr)
  module = runFastHeinz(network, scores)
  module
},mc.cores = n_cores)
module_lengths = sapply(module_list,function(module) length(V(module)))
module_lengths
num_nodes = as.numeric(params$num_nodes)
module_select = module_lengths <= num_nodes
module_select[length(module_select)] = TRUE
module_id = which(module_select)[1]
module = module_list[[module_id]]
module
```

Here the list of nodes within that module. Map gene id to gene name.

```{r}
gene_table = mapped[which(mapped$STRING_id %in% names(V(module))),]
head(gene_table)
```

Differential expression is coloured in red (upregulated) and green (downregulated). Scores are represented with shapes: rectangles are negative and circles are positive scores.

```{r fig.width=12, fig.height=12}
set.seed(1234)
node_symbol = sapply(names(V(module)),function(node_name) 
  gene_table$symbol[which(gene_table$STRING_id == node_name)])
if(length(V(module)) > 0) {
  par(mfrow=c(1,1))
  logFC = mapped$logFC
  names(logFC) = mapped$STRING_id
  plotModule(module, diff.expr = logFC, labels = node_symbol)
} else {
  message("no signficant subgraph found")
}
```

Interactive visualization.

```{r fig.width=12, fig.height=12}
if(length(V(module)) > 0) {
  V(module)$label = node_symbol
  visIgraph(module, idToLabel = FALSE)
} else {
  message("no signficant subgraph found")
}
```

Interactive visualization with physics animation.

```{r fig.width=12, fig.height=12}
if(length(V(module)) > 0) {
  V(module)$label = node_symbol
  visIgraph(module, idToLabel = FALSE, physics = TRUE)
} else {
  message("no signficant subgraph found")
}
```

## Compare to Known Pathways

Calculate number of genes that our network on known pathways have in common.

```{r}
# get human pathways form kegg database
pathways = keggList("pathway", "hsa")
human_pathways = sub("path:", "", names(pathways))
# kegg server only allow 10 request at the time
n_request = ceiling(length(human_pathways)/10)
chunk = function(x, n) split(x, sort(rank(x) %% n))
chunks_pathways = chunk(1:length(human_pathways),n_request)
# download from kegg server
list_chunk_pathways = lapply(chunks_pathways,function(one_chunk_pathways) {
  cat("download chunk:",one_chunk_pathways,"\n")
  pathway_ids = human_pathways[one_chunk_pathways]
  setNames(keggGet(pathway_ids), pathway_ids)
  })
# flatten list of lists
all_pathways = unlist(list_chunk_pathways, recursive=FALSE)
# check for all human pathways
gene_symbol_unique = unique(gene_table$symbol)
gene_overlap = sapply(all_pathways,function(pathway) {
  genes_one_pathway = pathway$GENE[c(FALSE, TRUE)]
  detected = sapply(genes_one_pathway,function(pathway_gene_str) {
    pathway_gene = (strsplit(pathway_gene_str,";") %>% unlist())[1]
    sum(str_detect(pathway_gene,gene_symbol_unique))
  })
  list(name=pathway$NAME,overlap=mean(detected))
}) %>% t
ordered_ids = order(unlist(gene_overlap[,"overlap"]),decreasing = TRUE)
head(gene_overlap[ordered_ids,],10)
```

## Session Info

```{r session_info}
sessionInfo()
```
